### PR TITLE
Add test suite for Store component

### PR DESCRIPTION
**Pull request by juztamau5**
## Description

This PR adds a test suite for the Store class, which connects to the wallets and performs transactions. I wrote this while testing the SFT purchase page, and decided to PR the store test suite separately while I work more on the purchase component.

The test suite renders the store component and digitally performs a "click" action on the form to initiate wallet connection. I swapped out web3modal for the truffle HDWallet when it's detected that we're running on CI infra vs. in the browser.

### Secrets

Uses the following GitHub Actions secrets, if set:

  * INFURA_API_KEY
  * PRIVATE_KEY

I use `dotenv-defaults` to manage environment variables, because it degrades gracefully both in the browser and in node. It pulls in values in the following order of preference:

* Environment variable set by CI runner
* Override specified in `.env` file (created by dev)
* Default specified in `.env.defaults` file (included in repo)

As a result, values can be provided on different levels with sensible defaults.

#### Jest version

Jest 26 introduces a bug, so Jest 25 is required. See:

* https://github.com/facebook/jest/issues/7780#issuecomment-705566017

#### jsdom bug

A patch is required for `@ethersproject/providers` because it doesn't check for null values. Null values aren't allowed by the spec, so omitting the check is proper, but `jest-environemt-jsdom` (or `jsdom` itself) has timers that return null instead of an open handle.

See:

  * https://github.com/facebook/jest/issues/7933
  * https://github.com/bugsnag/bugsnag-js/issues/452

## How has this been tested?

Result of CI test:

```
|    $ react-scripts test --passWithNoTests
|    PASS src/__tests__/test-store.tsx (24.726 s)
|      ● Console
|
|        console.log
|          EventProvider launched on network:  rinkeby
|
|          at Store._launchEventProvider (src/stores/store.tsx:319:17)
```

We see the store component being rendered and submitted, and as a result the store connects to the network and logs its success: `EventProvider launched on network:  rinkeby`.

I had to insert large artificial pauses, because otherwise the test suite ends while web3 stuff is still going on in the background and it freezes the test runner. Adding the large pauses let background tasks settle before exiting, preventing the crashes. However, the web3 dependencies should be scrutinized to avoid this hack in the future.

### CODE REVIEW

**Comment by juztamau5** [comment]
Upon testing, it seems that the web3 stuff is hanging the test suite about 1/3 of the time. I'll try to fix this before landing the SFT pull request. However, because we run 3 tests in parallel, there's only 1/9 chance that all will hang, so not the end of the world if we merge this before the underlying bug is fixed.

**Comment by juztamau5** [comment]
@peak3d I got CI to pass successfully now.

For security our GitHub Actions secrets aren't exposed to PRs from external repos, so I inserted my Infura API key for a default value. After merging I'll upload a team API key that'll be used for internal runs not triggered from external PRs.

**Comment by peak3d** [comment]
@juztamau5
Regarding the wait / web3js: Wouldn't it be possible to add the ganache node module, spawn it for the test and use the new private network option to test on local node?
We will have to do this also in the next time to be able to fast forward time to test the staking reward contracts.
Reference: uniswap liquidity-staking repo.

File: .env.defaults [hunk 1]
```diff
@@ -0,0 +1,29 @@
...
+# Account Details > Export Private Key.
+#
+# Be aware of NEVER putting real Ether into testing accounts.
+#
+# Ropsten faucet: https://faucet.metamask.io/
+# Rinkeby faucet: https://faucet.rinkeby.io/
+# Goerli faucet: https://goerli-faucet.slock.it/
+# Kovan faucet: https://faucet.kovan.network/
+#
+PRIVATE_KEY=0xc87509a1c067bbde78beb793e6fa76530b6382a4c0241e5e4a9ec0a0f44dc0d3
```
Comment (peak3d) [inline .env.defaults]: what is this private key?
From what I know env files should never be part of the public git repo.

File: .github/workflows/node.js.yml [hunk 2]
```diff
@@ -6,6 +6,12 @@
 #
 #   https://help.github.com/actions/language-and-framework-guides/using-nodejs-with-github-actions
 #
+# This workflow requires the following GitHub Actions secrets:
+#
+#   * INFURA_API_KEY: Your API key for Infura. Obtain one at https://infura.io/
+#   * PRIVATE_KEY: The wallet private key or mnemonic for determing the accounts
+#     to use for transactions
+#
 ################################################################################
```
Comment (peak3d) [inline .github/workflows/node.js.yml:15]: We should go with node ganache module and spawn our own eth node.
Then we can also remove the dotenv file in this git repo.

Having a private key and our private infura id in public git repo is IMO not the way to go.

File: src/stores/store.tsx [hunk 3]
```diff
@@ -136,17 +138,37 @@ class Store {
...
+      // provider
+      if (PRIVATE_KEY !== undefined) {
+        if (!INFURA_API_KEY) {
+          throw new Error('INFURA_API_KEY is required');
+        }
+
+        const web3Provider = new HDWalletProvider(
+          PRIVATE_KEY,
+          `https://${this.networkName}.infura.io/v3/${INFURA_API_KEY}`
+        );
```
Comment (peak3d) [inline src/stores/store.tsx:155]: I don't see the requirement adding truffle code here.
In ethers you create a "Wallet" which acts as the signer.

File: src/stores/store.tsx [hunk 4]
```diff
@@ -136,17 +138,37 @@ class Store {
         await this.disconnect(false);
       }
 
-      let ethersProvider: ethers.providers.JsonRpcProvider;
-      if (this.networkName === 'private') {
+      let ethersProvider:
+        | ethers.providers.JsonRpcProvider
+        | ethers.providers.Web3Provider;
```
Comment (peak3d) [inline src/stores/store.tsx:143]: Web3Provider inherits from JsonRpcProvider, where do you need special access on web3Provider properties?

**Review by peak3d — CHANGES_REQUESTED** [review]

**Comment by peak3d** [comment]
@juztamau5 sure we should add the possibility for HDWallets in the future, but this needs more work, for example listing all addresses. For the CI run I really want to suggest that we use local ganache, which is spawned on CI runtime.

It offers beside the fact that you don't have to implement HDWallet additionally great features like time fast-forwarding.

**Comment by juztamau5** [comment]
I'm going to close this. Given that @peak3d  is focusing on UI and I'm focusing on SOL, I'd like to not mingle them right now to give us both enhanced speed. It's not worth polluting UI with build system stuff

**Pull request closed without merging**